src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java

```diff
@@ -2211,8 +2215,12 @@ public int read(byte[] b, int off, int len) throws IOException {
             ByteBuffer bb = ByteBuffer.wrap(b);
             bb.position(off);
             bb.limit(off + len);
-            synchronized(zfch) {
-                n = zfch.position(pos).read(bb);
+            if (zfch instanceof FileChannel) {
+                n = ((FileChannel) zfch).read(bb, pos);
+            } else {
+                synchronized (zfch) {
+                    n = zfch.position(pos).read(bb);
+                }
```

Comment:
@LanceAndersen Are you planning to look at this? Do you mind checking the async close case to make sure that the synchronization isn't masking anything?

Also just to point out that pattern matching for instanceof ca be used here.

Reply:
Yes, I plan to look at this.  It would also be good to have a couple of additional reviews as well :-)

Reply:
I think using the positional read on the underlying FileChannel is okay. I'm puzzled by the previous code as I would have expected it to restore the position.

Reply:
My reading of the existing code is that the only position-influenced method called on the channel (either via `ZipFileSystem.ch` or `ZipFileSystem$EntryInputStream.zfch`) is `read`, and this is only called in the `.position(pos).read(...)` idiom. The failure to reset the position doesn't affect correctness. However the `synchronzized` _is_ definitely needed to avoid races.

Incidentally, regarding this comment:

    private class EntryInputStream extends InputStream {
        private final SeekableByteChannel zfch; // local ref to zipfs's "ch". zipfs.ch might
                                                // point to a new channel after sync()

If the file system is writable and updated, the underlying file is deleted and replaced with a temporary file by `close()` / `sync()`, but `ZipFileSystem.ch` is itself final since d581e4f4. I believe the comment is outdated and `EntryInputStream` could just access ch via the outer pointer. That change would simplify this patch marginally.

Reply:
I've added the simplifying commit for now, but I'm happy to split that to a separate change if you prefer.